### PR TITLE
started crud implementation

### PR DESCRIPTION
**Pull request by aalvarez11**
Began the CRUD implementation, starting with checking if code was able to retrieve dummy rows added to the database. 

`INSERT INTO product(id, name, quantity, price) VALUES (gen_random_uuid(), 'elote', 26, 2.00);`
`INSERT INTO product(id, name, quantity, price) VALUES (gen_random_uuid(), 'conchas', 30, 1.00);`

Screen Shot 2022-08-12 at 4 43 47 PM

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,22 @@
...
+
+To view and access the database in the terminal using bash, use the command `docker exec -it pass-bakery-postgres bash` from there use the bash command `psql -U user` If prompted for the password, provide `pass` or whatever password you made during the `docker run` command.
+
+Create the database used in this project with psql command `create database pass_bakery_db` You can check your local databases using `\l` to make sure pass_bakery_db now exists in your databases. Next, you can connect to it with `\c pass_bakery_db` and you are now able to execute sql in the terminal, like making tables and populating them with records preemptively or for testing.
+
+To log back into the db at any time, run this command in the terminal: `psql -h localhost -p 5432 -U user` afterwards you will be prompted for the password, and it is simply `pass` or any custom password you used in setup.
```
Comment (fsasieta) [inline README.md]: You are mixing a docker cheat sheet with instructions on running the application. Separate the to differently purposed sets of information. For running the application, keep it as a bullet list, i.e. 1) install if not installed, 2) run database, 3) run the application. Notice you don't say "run `sbt run` anywhere, so you missed that as well. The additional info regarding the database is helpful though, so keep that in a different section

File: README.md [hunk 2]
```diff
@@ -1,2 +1,22 @@
-# pass-bakery
-scala training project
+# Pass-Bakery
+Scala training project
+
+### Project Dependencies
```
Comment (fsasieta) [inline README.md]: You title is one #, the next titles should be 2 ##, to keep the semantic meaning of the document

File: app/DAO/BakeryDatabase.scala [hunk 3]
```diff
@@ -35,4 +35,77 @@ class BakeryDatabase @Inject() (
...
+            ) + ", price: " + selectResult.getDouble(
+              "price"
+            ) + ", created at: " + selectResult.getTimestamp(
+              "created_at"
+            ) + ", updated at: " + selectResult.getTimestamp(
+              "updated_at"
+            ) + "\n"
+            idx += 1
+          }
+          returnStr
```
Comment (fsasieta) [inline app/DAO/BakeryDatabase.scala:63]: I sometimes forget what dealing with individual raw rows looks like. I think this is about time you integrate with the doobie library, which should make things a bit nicer to deal with

**Review by fsasieta — COMMENTED** [review]
You are missing unit tests

File: README.md [hunk 4]
```diff
@@ -1,2 +1,24 @@
...
+* play json
+* jdbc (used by Scala in the Play Framework)
+* evolutions
+* postgres
+* doobie
+
+## Running the Application
+1. Install [Docker](https://www.docker.com/products/docker-desktop/) and  [postgresql](https://www.postgresql.org/download/) on your machine, if not installed already.
+2. Launch Docker (this app expects the latest postgres alpine)
+3. Run the database through terminal commands
```
Comment (fsasieta) [inline README.md:14]:
```suggestion
3. Run the database through terminal commands
    1. Run the database with the command:
    2. docker run --name pass-bakery-postgres -e POSTGRES_USER=user -e POSTGRES_PASSWORD=pass -p 5432:5432 -d postgres:alpine
```
You expect a specific user, database name and password as well as port. Do not make your prospective users or future developer guess how to do their setup.

You mention one time things below. Do these hold if someone stops the docker container? If not, they are not one time things, they need to be done everytime someone starts the database

File: app/controllers/APIController.scala [hunk 5]
```diff
@@ -35,4 +35,22 @@ class APIController @Inject() (
         Ok(databaseTables)
       }
   }
+
+//  def createProduct() = Action.async { implicit request: Request[AnyContent] =>
+//    Ok("new item added")
+//  }
```
Comment (fsasieta) [inline app/controllers/APIController.scala]: Remove

File: app/controllers/APIController.scala [hunk 6]
```diff
@@ -35,4 +35,22 @@ class APIController @Inject() (
...
+
+  def getProduct(id: String) = Action.async {
+    implicit request: Request[AnyContent] =>
+      bakeryDB.getProductById(id).map {
+        case ""                => NotFound("No Product Found")
+        case myProduct @ found => Ok(myProduct.toString)
+      }
```
Comment (fsasieta) [inline app/controllers/APIController.scala:44]: The preferred way in which we will treat emptyness in scala (as opposed to traditional java; not the modern one) will be with `Option`. To take advantage of them, you have to define your function to return a `Future[Option[String]]`

File: app/DAO/BakeryDatabase.scala [hunk 7]
```diff
@@ -35,4 +35,97 @@ class BakeryDatabase @Inject() (
...
+        try {
+          val statement = {
+            conn.prepareStatement("SELECT * FROM product WHERE id::text = ?")
+          }
+          statement.setString(1, id)
+
+          println(statement.toString)
+          val selectResult = statement.executeQuery()
+          var returnStr = ""
+          while (selectResult.next()) {
```
Comment (fsasieta) [inline app/DAO/BakeryDatabase.scala]: Do you need a while loop here? you really are only getting one element right? (can you be sure it is unique?)

Comment (aalvarez11) [reply app/DAO/BakeryDatabase.scala]: the while loop here is to iterate over the row's columns

Comment (fsasieta) [reply app/DAO/BakeryDatabase.scala]: resolved while on our meeting

File: app/DAO/BakeryDatabase.scala [hunk 8]
```diff
@@ -35,4 +35,100 @@ class BakeryDatabase @Inject() (
...
+            conn.prepareStatement("INSERT INTO product() VALUES ()")
+          val insertResult = statement.executeQuery()
+        } catch {
+          case e: Exception => e.printStackTrace().toString
+        }
+
+        t
+      }
+    }
+  }
```
Comment (fsasieta) [inline app/DAO/BakeryDatabase.scala]: This is also dead code/ not getting used. Remove from this PR

File: app/DAO/BakeryDatabase.scala [hunk 9]
```diff
@@ -35,4 +35,100 @@ class BakeryDatabase @Inject() (
...
+                "updated_at"
+              )
+            )
+          } else {
+            None
+          }
+          result
+        } catch {
+          case e: Exception => Option(e.printStackTrace().toString)
+        }
```
Comment (fsasieta) [inline app/DAO/BakeryDatabase.scala:104]: Read about `Try()`, which is a Scala construct that you will see often. No action to take on this so far. Some links:

https://alvinalexander.com/scala/try-success-failure-example-reading-file/
https://medium.com/@shantanukamath/why-try-scala-error-handling-d71d2bd44d8c

The advantage of this over try-catch is that sometimes reading error handling is easier (with match case statements). Similar to what we did with options

File: conf/routes [hunk 10]
```diff
@@ -11,5 +11,12 @@ GET     /hello                      controllers.HomeController.hello()
...
 
 # API endpoints
-GET     /pass-bakery/status         controllers.APIController.getStatus()
-GET     /pass-bakery/db-tables      controllers.APIController.getDatabaseTables()
+GET      /pass-bakery/status         controllers.APIController.getStatus()
+GET      /pass-bakery/db-tables      controllers.APIController.getDatabaseTables()
+#POST    /rest/bakery/product        controllers.APIController.createProduct(prod: JsValue)
+GET     /rest/bakery/product/:id    controllers.APIController.getProduct(id: String)
+#PUT     /rest/bakery/product/:id    controllers.APIController.updateProduct(id: String, prod: JsValue)
+#DELETE  /rest/bakery/product/:id    controllers.APIController.deleteProduct(id: String)
```
Comment (fsasieta) [inline conf/routes]: Remove commented out endpoints

**Review by fsasieta — APPROVED** [review]
This is a good checkpoint. Lets remove the comments and merge

**Pull request merged**